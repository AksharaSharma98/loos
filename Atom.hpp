/*
  Atom.hpp
  (c) 2008 Tod D. Romo


  Grossfield Lab
  Department of Biochemistry and Biophysics
  University of Rochester Medical School

  Basic atom class
*/


#if !defined(ATOM_HPP)
#define ATOM_HPP


#include <iostream>
#include <string>
#include <stdexcept>
#include <vector>
#include <algorithm>
#include <tr1/memory>


using namespace std;
using namespace tr1;

#include <loos.hpp>

class Atom;

//! Shared pointer to an Atom
typedef shared_ptr<Atom> pAtom;

//! Basic Atom class for handling atom properties.
/*!
  This class handles atoms and atom properties.  It stores a GCoord
  coordinate internally.  Bonds are included, but are represented as a
  vector of atom-id's, which are assumed to be unique per atom...

  Most properties are derived from the PDB file specification.
  Exceptions are noted below.  Accessors for each property are
  provided and should be self-explanatory...
*/

  
class Atom {
public:
  Atom() { init(); }

  //! Constructs an atom with the atomid i, atomname s, and coordinates c.
  /*!
    Constructs a new atom.
    \param i atom-id
    \param s atom-name
    \param c Coordinates
  */

  Atom(const int i, const string s, const GCoord& c) {
    init();
    _id = i;
    _name = s;
    _coords = c;
  }


  // Accessors...
  int id(void) const { return(_id); }
  void id(const int i) { _id = i; }
  
  int resid(void) const { return(_resid); }
  void resid(const int i) { _resid = i; }

  string name(void) const { return(_name); }
  void name(const string s) { _name = s; }

  string altLoc(void) const { return(_altloc); }
  void altLoc(const string s) { _altloc = s; }

  string chainId(void) const { return(_chainid); }
  void chainId(const string s) { _chainid = s; }

  string resname(void) const { return(_resname); }
  void resname(const string s) { _resname = s; }

  string segid(void) const { return(_segid); }
  void segid(const string s) { _segid = s; }

  string iCode(void) const { return(_icode); }
  void iCode(const string s) { _icode = s; }

  string PDBelement(void) const { return(_pdbelement); }
  void PDBelement(const string s) { _pdbelement = s; }

  //! Returns a const ref to internally stored coordinates.
  //! This returns a const ref mainly for efficiency, rather than
  //! copying the coords...
  const GCoord& coords(void) const { return(_coords); }
  void coords(const GCoord& c) { _coords = c; }

  double bfactor(void) const { return(_b); }
  void bfactor(const double d) { _b = d; }

  double occupancy(void) const { return(_q); }
  void occupancy(const double d) { _q = d ; }

  double charge(void) const { return(_charge); }

  //! Sets the charge of the atom as a double.  This is NOT the PDB spec...

  void charge(const double d) { _charge = d ; }

  double mass(void) const { return(_mass); }
  void mass(const double d) { _mass = d ; }

  //! Recordname imported from the PDB for this Atom
  //! This is mainly for atoms that come from a PDB, i.e. whether or
  //! not they were an ATOM or a HETATM
  string recordName(void) const { return(_record); }
  void recordName(const string s) { _record = s; }

  //! Clear all stored bonds
  void clearBonds(void) { bonds.clear(); }
  //! Add a bond given a pAtom (extracting the atomid of the bond)
  void addBond(const pAtom& p) { bonds.push_back(p->id()); }
  //! Add a bond to an atom-id
  void addBond(const int i) { bonds.push_back(i); }

  //! Deletes the specified bond.
  void deleteBond(const int b) {
    vector<int>::iterator i = find(bonds.begin(), bonds.end(), b);
    if (i == bonds.end())
      throw(runtime_error("Attempting to delete a non-existent bond"));
    bonds.erase(i);
  }

  //! Deletes a bond by extracting the atom-id from the passed pAtom
  void deleteBond(const pAtom& p) { deleteBond(p->id()); }

  //! Returns a copy of the bond list.
  vector<int> getBonds(void) const { return(bonds); }

  bool hasBonds(void) const { return(bonds.size() != 0); }


  //! Outputs an atom in pseudo-XML
  friend ostream& operator<<(ostream& os, const Atom& a) {
    os << "<ATOM ID='" << a._id << "' NAME='" << a._name << "' ";
    os << "RESID='" << a._resid << "' RESNAME='" << a._resname << "' ";
    os << "COORDS='" << a._coords << "' ";
    os << "ALTLOC='" << a._altloc << "' CHAINID='" << a._chainid << "' ICODE='" << a._icode << "' SEGID='" << a._segid << "' ";
    os << "B='" << a._b << "' Q='" << a._q << "' CHARGE='" << a._charge << "' MASS='" << a._mass << "'";
    if (a.hasBonds() > 0) {
      vector<int>::const_iterator i;
      os << ">\n";
      for (i=a.bonds.begin(); i != a.bonds.end(); i++)
	os << "  <BOND>" << *i << "</BOND>\n";
      os << "</ATOM>";
    } else
      os << "/>";
    return(os);
  }

private:
  void init() {
    _id = -1;
    _resid = -1;
    _b = _q = 0.0;
    _charge = 0.0;
    _mass = 1.0;
    _name = "    ";
    _altloc = " ";
    _resname = "   ";
    _chainid = " ";
    _segid = "    ";
    _pdbelement = "";
    _record = "ATOM";
  }



private:
  int _id;
  string _record, _name, _altloc, _resname, _chainid;
  int _resid;
  string _icode;
  double _b, _q, _charge, _mass;
  string _segid, _pdbelement;
  GCoord _coords;

  vector<int> bonds;
};


#endif
