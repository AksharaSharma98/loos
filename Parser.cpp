/*
  Parser.hpp
  (c) 2008 Tod D. Romo


  Grossfield Lab
  Department of Biochemistry and Biophysics
  University of Rochester Medical School

  Parser for atom selections...
*/

#include <sstream>

#include "Parser.hpp"

using namespace std;


#define GPTOKS(x)    (boost::get<0>((x)))
#define GLTOKS(x)    (boost::get<1>((x)))

namespace loos {

  namespace parser {

    void dump(const pares& p, const string s) {
      cout << s << endl;
      ptoks q = GPTOKS(p);
      Tokens r = GLTOKS(p);

      cout << "  o Parse tokens o\n";
      prptoks(q);
      cout << "  o Remaining lex tokens o\n";
      cout << r;
    }

    int max3(const int x, const int y, const int z) {
      int d = (x > y) ? x : y;
      d = (d > z) ? d : z;
      return(d);
    }

    ptok mktok(const ptok::ptok_type t, const string s = "") {
      ptok p;

      p.type = t;
      p.val = s;
      return(p);
    }

    void prptoks(const ptoks& p) {
      ptoks::const_iterator i;
      for (i = p.begin(); i != p.end(); i++)
	cout << (*i) << endl;
    }

    pares mknull(const pares& p) {
      Tokens t = GLTOKS(p);
      ptoks r;
      pares q(r, t);
      return(q);
    }


    // p may ref a temp...  Problem???
    pares mkpares(const ptok& p, const Tokens& t) {
      ptoks r;
      r.push_back(p);
      pares q(r, t);
      return(q);
    }

    int ptokn(const pares& p) {
      ptoks r = GPTOKS(p);
      return(r.size());
    }

    int depth(pares *p, pares *q=0, pares *r=0) {
      int d = ptokn(*p);
      if (d == 0)
	return(0);

      if (q != 0) {
	int e = ptokn(*q);
	if (e == 0)
	  return(0);
	d += e;

	if (r != 0) {
	  e = ptokn(*r);
	  if (e == 0)
	    return(0);
	  d += e;
	}
      }
      return(d);
    }

    ptok gptok(const pares& p) {
      ptoks r = GPTOKS(p);
      return(r.front());
    }


    // p : q  | q
    pares twoway(const pares& p, const pares& q) {
      ptoks pt = GPTOKS(p);
      ptoks qt = GPTOKS(q);
      Tokens lt = GLTOKS(q);

      qt.insert(qt.end(), pt.begin(), pt.end());
      pares res(pt, lt);
      return(res);
    }

    // p : r : q
    pares threeway(const pares& p, const pares& q, const pares& r) {
      ptoks pt = GPTOKS(p);
      ptoks qt = GPTOKS(q);
      ptoks rt = GPTOKS(r);
      Tokens lt = GLTOKS(r);

      pt.insert(pt.end(), rt.begin(), rt.end());
      pt.insert(pt.end(), qt.begin(), qt.end());
      pares res(pt, lt);
      return(res);
    }

    pares zap(const pares& p) {
      ptoks notoks;
      Tokens toks = GLTOKS(p);
      pares res(notoks, toks);
      return(res);
    }



    pares numid(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);


      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::ID) {
	ptok p;

	if (t.datum == "id")
	  p = mktok(ptok::id);
	else if (t.datum == "resid")
	  p = mktok(ptok::resid);
	else
	  return(null);

	return(mkpares(p, ltoks));
      }

      return(null);
    }



    pares alphid(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::ID) {
	ptok p;

	if (t.datum == "name")
	  p = mktok(ptok::name);
	else if (t.datum == "resname")
	  p = mktok(ptok::resname);
	else if (t.datum == "segid")
	  p = mktok(ptok::segid);
	else
	  return(null);

	return(mkpares(p, ltoks));
      }

      return(null);
    }


    pares number(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::NUMERIC) {
	ptok p = mktok(ptok::num, t.datum);
	return(mkpares(p, ltoks));
      }

      
      return(null);
    }



    pares alpha(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::STRING) {
	ptok p = mktok(ptok::alpha, t.datum);
	return(mkpares(p, ltoks));
      }

      
      return(null);
    }



    pares numeric(const pares p) {

      pares p1 = numid(p);
      pares p2 = number(p);

      if (ptokn(p1) >= ptokn(p2))
	return(p1);
      
      return(p2);
    }




    pares alphabetic(const pares p) {

      pares p1 = alphid(p);
      pares p2 = alpha(p);

      if (ptokn(p1) >= ptokn(p2))
	return(p1);
      
      return(p2);
    }


    pares literal(const pares p) {

      pares p1 = alphabetic(p);
      pares p2 = numeric(p);

      if (ptokn(p1) >= ptokn(p2))
	return(p1);
      
      return(p2);
    }

    pares unop(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::OPERATOR) {
	ptok p;
	if (t.datum == "!")
	  p = mktok(ptok::lnot);
	else
	  return(null);
	return(mkpares(p, ltoks));
      }
      return(null);
    }


    pares relop(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::OPERATOR) {
	ptok p;
	if (t.datum == "<")
	  p = mktok(ptok::lt);
	else if (t.datum == "<=")
	  p = mktok(ptok::lte);
	else if (t.datum == ">=")
	  p = mktok(ptok::gte);
	else if (t.datum == ">")
	  p = mktok(ptok::gt);
	else if (t.datum == "==")
	  p = mktok(ptok::eq);
	else if (t.datum == "!=")
	  p = mktok(ptok::neq);
	else
	  return(null);
	return(mkpares(p, ltoks));
      }
      return(null);
    }


    pares binop(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::OPERATOR) {
	ptok p;
	if (t.datum == "&&")
	  p = mktok(ptok::land);
	else if (t.datum == "||")
	  p = mktok(ptok::lor);
	else
	  return(null);
	return(mkpares(p, ltoks));
      }
      return(null);
    }

    pares regexp(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      pares p1 = alphid(p);
      if (ptokn(p1) == 0)
	return(null);

      ltoks = GLTOKS(p1);
      Token t = ltoks.pop();
      if (t.type == Token::OPERATOR && t.datum == "=~") {
	ptoks notoks;
	pares p2 = alpha(pares(notoks, ltoks));
	if (ptokn(p2) > 0) {
	  pares q = mkpares(mktok(ptok::regex), ltoks);
	  return(threeway(p1, q, p2));
	}
      }
      return(null);
    }


    pares rexp(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      pares p1 = regexp(p);
      
      pares p2a = literal(p);
      pares p2b = relop(zap(p2a));
      pares p2c = literal(zap(p2b));

      if (ptokn(p1) > 0)
	return(p1);
      if (ptokn(p2a) >0 && ptokn(p2b) >0 && ptokn(p2c) >0)
	return(threeway(p2a, p2b, p2c));
      
      return(null);
    }


    pares lexp(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

    
      pares p1 = rexp(zap(p));

      pares p2a = unop(zap(p));
      pares p2b = rexp(zap(p2a));

      pares p3a = rexp(zap(p));
      pares p3b = binop(zap(p3a));
      pares p3c = rexp(zap(p3b));

      int d1 = depth(&p1);
      int d2 = depth(&p2a, &p2b);
      int d3 = depth(&p3a, &p3b, &p3c);

      int m = max3(d1, d2, d3);
      
      if (m == d1)
	return(p1);
      else if (m == d2)
	return(twoway(p2a, p2b));

      return(threeway(p3a, p3b, p3c));

    }


    pares sexp(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

      Token t = ltoks.pop();
      if (t.type == Token::LPAR) {
	ptoks notoks;
	pares q(notoks, ltoks);
	pares r = lexp(q);
	if (ptokn(r) > 0) {
	  ltoks = GLTOKS(r);
	  t = ltoks.pop();
	  if (t.type == Token::RPAR) {
	    ptoks toks = GPTOKS(r);
	    pares res(toks, ltoks);
	    return(res);
	  }
	}
      }
      
      return(lexp(p));
    }


    pares expr(const pares p) {
      Tokens ltoks = GLTOKS(p);
      pares null = mknull(p);

      if (ltoks.empty())
	return(null);

    
      pares p1 = sexp(zap(p));

      pares p2a = unop(zap(p));
      pares p2b = sexp(zap(p2a));

      pares p3a = sexp(zap(p));
      pares p3b = binop(zap(p3a));
      pares p3c = sexp(zap(p3b));

      int d1 = depth(&p1);
      int d2 = depth(&p2a, &p2b);
      int d3 = depth(&p3a, &p3b, &p3c);

      int m = max3(d1, d2, d3);
      
      if (m == d1)
	return(p1);
      else if (m == d2)
	return(twoway(p2a, p2b));

      return(threeway(p3a, p3b, p3c));

    }



    
  };

};


