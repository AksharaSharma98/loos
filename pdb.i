/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/


%header %{
#include <pdb.hpp>
%}


namespace loos {
  class PDB : public AtomicGroup {
  public:
    PDB();
    virtual ~PDB();
    explicit PDB(const std::string fname);
    explicit PDB(const char* fname);
    virtual PDB* clone(void) const;
    PDB copy(void) const;
    static PDB fromAtomicGroup(const AtomicGroup&);
    void showCharge(bool b = true);
    void strict(const bool b);
    void autoTerminate(bool b = true);
    Remarks& remarks(void);
    void remarks(const Remarks&);
    const UnitCell& unitCell(void);
    void unitCell(const UnitCell&);

    %extend {
      char* __str__() {
        std::ostringstream oss;
        oss << *$self;
        size_t n = oss.str().size();
        char* buf = new char[n+1];
        strncpy(buf, oss.str().c_str(), n+1);
        return(buf);
      }
    }

  };

}
