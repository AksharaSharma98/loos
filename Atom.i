%module Atom
%include "std_string.i"
%include "std_vector.i"
%include <boost_shared_ptr.i>

%template(IntVector) std::vector<int>;
%shared_ptr(loos::Atom)

%include "Coord.i"

%{
#include <iostream>
#include <string>
#include <stdexcept>
#include <vector>
#include <functional>

#include <boost/shared_ptr.hpp>
#include <boost/tuple/tuple.hpp>

#include <loos_defs.hpp>
#include "Atom.hpp"

%}




namespace loos {

  class Atom;
  typedef Coord<double>   GCoord;
  typedef boost::shared_ptr<Atom> pAtom;


  //! Basic Atom class for handling atom properties.
  /**
   * This class handles atoms and atom properties.  It stores a GCoord
   * coordinate internally.  Bonds are included, but are represented as a
   * vector of atom-id's, which are assumed to be unique per atom...
   *
   * Most properties are derived from the PDB file specification.
   * Exceptions are noted below.  Accessors for each property are
   * provided and should be self-explanatory...
   */

  
  class Atom {
  public:

    // Be careful that we don't overflow an unsigned long!
    //! Bits in the bitmask that flag what properties have actually been set.
    enum bits {
      nullbit = 0,
      coordsbit = 1,
      bondsbit = coordsbit << 1,
      massbit = bondsbit << 1,
      chargebit = massbit << 1,
      anumbit = chargebit << 1,
      flagbit = anumbit << 1,
      usr1bit = flagbit << 1,
      usr2bit = usr1bit << 1,
      usr3bit = usr2bit << 1
    };

    Atom() { init(); }
    Atom(const int i, const std::string s, const GCoord& c) {
      init();
      _id = i;
      _name = s;
      _coords = c;
    }
    ~Atom() { }

    // Accessors...
    int id(void) const;
    void id(const int);
  
    int resid(void) const;
    void resid(const int);

    int atomic_number(void) const;
    void atomic_number(const int);

    std::string name(void) const;
    void name(const std::string);

    std::string altLoc(void) const;
    void altLoc(const std::string);

    std::string chainId(void) const;
    void chainId(const std::string);

    std::string resname(void) const;
    void resname(const std::string);

    std::string segid(void) const;
    void segid(const std::string);

    std::string iCode(void) const;
    void iCode(const std::string);

    std::string PDBelement(void) const;
    void PDBelement(const std::string);

    const GCoord& coords(void) const;

    GCoord& coords(void);

    void coords(const GCoord&);

    double bfactor(void) const;
    void bfactor(const double);

    double occupancy(void) const;
    void occupancy(const double);

    double charge(void) const;

    void charge(const double);

    double mass(void) const;
    void mass(const double);

    std::string recordName(void) const;
    void recordName(const std::string);

    void clearBonds(void);

    void addBond(const pAtom&);

    void addBond(const int);


    void deleteBond(const int);

    void deleteBond(const pAtom&);

    std::vector<int> getBonds(void) const;

    void setBonds(const std::vector<int>& list);

    bool hasBonds(void) const;

    //! Checks to see if this atom is bound to another atom
    bool isBoundTo(const int);

    bool isBoundTo(const pAtom&);

    //! Given a bit-mask, checks to see if those bits are set.
    /** For example, to check whether or not the coords have been set,
     *  do,
     \verbatim
     checkProperty(Atom::coordsbit)
     \endverbatim
     *  You can combine checks by or'ing the bit flags, such as the
     *  following which checks whether both mass and charge have been
     *  set,
     \verbatim
     checkProperty(Atom::massbit | Atom::chargebit)
     \endverbatim
    */
    bool checkProperty(const bits bitmask);

    
    //! Sets user-defined bits
    /** The user-available property bits are flagbit, usr1bit,
     * usr2bit, and usr3bit.  Attempting to set any other bits will
     * cause a logic_error exception to be thrown.
     *
     * This facility is useful if you want to tag atoms for later
     * processing without putting them in a separate group, for
     * example...
     */
    void setProperty(const bits bitmask);
    

    //! Clears user-defined bits...
    void clearProperty(const bits bitmask);

    //! Outputs an atom in pseudo-XML
    //friend std::ostream& operator<<(std::ostream&, const Atom&);

  private:
    void init(void);
    //! Internal function for setting a bitflag
    void setPropertyBit(const bits);
    //! Internal function for clearing a bitflag
    void clearPropertyBit(const bits);

    void checkUserBits(const bits bitmask);

  private:
    int _id;
    std::string _record, _name, _altloc, _resname, _chainid;
    int _resid;
    int _atomic_number;
    std::string _icode;
    double _b, _q, _charge, _mass;
    std::string _segid, _pdbelement;
    GCoord _coords;
    unsigned long mask;

    std::vector<int> bonds;
  };



  // //! Compares two atoms based solely on name, id, resid, resname, and segid
  // struct AtomEquals : public std::binary_function<pAtom, pAtom, bool> {
  //   bool operator()(const pAtom& a, const pAtom& b) const;
  // };


  // //! Compares two atoms based on name, id, resid, resname, segid, and coords
  // /**
  //  * The default distance threshold is 1e-3 Angstroms
  //  */
  // struct AtomCoordsEquals : public std::binary_function<pAtom, pAtom, bool> {
  //   AtomCoordsEquals(const double d) : threshold(d*d) { }
  //   AtomCoordsEquals() : threshold(1e-6) { }

  //   bool operator()(const pAtom& a, const pAtom& b) const;
  //   double threshold;
  // };


};

