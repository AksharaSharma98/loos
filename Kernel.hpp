/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/





#if !defined(KERNEL_HPP)
#define KERNEL_HPP

#include <vector>

#include "Atom.hpp"

#include "KernelValue.hpp"
#include "KernelStack.hpp"
#include "KernelActions.hpp"

using namespace std;

namespace loos {

  //!The Kernel (virtual machine) for compiling and executing user-defined atom selections

  class Kernel {
    vector<Action*> actions;    //! Commands
    ValueStack val_stack;       //! The data stack...

  public:
    
    //! Destroy all of the stored commands...
    ~Kernel() {
      vector<Action*>::iterator i;
      for (i = actions.begin(); i != actions.end(); i++)
	delete (*i);
    }

    //! Add a command, setting the data-stack pointer...
    void push(Action *act) {
      act->setStack(&val_stack);
      actions.push_back(act); 
    }

    void pop(void) { actions.pop_back(); }

    //! Execute the stored commands for a specific atom.
    //! This does not clear the command stack/list
    void execute(pAtom pa = pAtom()) {

      vector<Action*>::iterator i;
      for (i=actions.begin(); i != actions.end(); i++) {
	(*i)->setAtom(pa);
	(*i)->execute();
      }
    }

    
    void clearActions(void) { actions.clear(); }

    ValueStack& stack(void) { return(val_stack); }

    friend ostream& operator<<(ostream& os, const Kernel& k) {
      vector<Action*>::const_iterator i;

      os << "Commands:\n";
      for (i=k.actions.begin(); i != k.actions.end(); i++)
	os << (*i)->name() << endl;
      os << endl;
      return(os);
    }

  };

};


#endif
