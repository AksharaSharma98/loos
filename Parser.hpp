/*
  Parser.hpp
  (c) 2008 Tod D. Romo


  Grossfield Lab
  Department of Biochemistry and Biophysics
  University of Rochester Medical School

  Parser for atom selections using Bison/Flex

*/


#if !defined(PARSER_HPP)
#define PARSER_HPP

#include <iostream>
#include <sstream>
#include <stdexcept>

using namespace std;

#include <AtomicGroup.hpp>
#include <Kernel.hpp>
#include <ParserDriver.hpp>


//! Front-end to the Bison/Flex parser
/** Creates a compiled Kernel that can then be executed to select
 *  atoms.  The grammar accepted is relatively simple and patterned
 *  after C/PERL expressions.  Relational operators are allowed, as
 *  are basic logical operators (and, or, and not).  Pre-defined
 *  keywords are: name, id, resname, resid, segid.  These evaluate to
 *  the current atom's appropriate property.  Case IS significant both
 *  for keywords and for strings.  Integer numbers are allowed.
 *  Strings are delimited by either single quotes (') or double quotes
 *  (").  Also, strings inequalities are handled lexically as C++
 *  strings are, i.e. string1 > string2 is true is it would be in
 *  regular C++ code.
 *
 *  Regular expressions (in PERL format) are supported via Boost.  The
 *  regular expression matching operator, '=~' is slightly special in
 *  that it will only permit you to match a keyword that would
 *  evaluate to a string.  In other words, you may match against a
 *  name, resname, and segid, but NOT an id nor a resid.
 *
 *  String equality matches the entire string.  If you want to match a
 *  subset, you should use the '=~' operator.  In other words,
 *  \verbatim
 *    "CA" == "C"   -> false
 *    "C"  == "C"   -> true
 *    "CA" =~ "C"   -> true
 *  \endverbatim
 *
 *  Finally, the standard precedence and associativity that apply in
 *  C++ apply here.  Expressions are evaluated left to right and
 *  parenthesis may be used to alter precedence/evaluation order.
 *  Unlike C/C++ however, the logical operators do not short-circuit.
 *
 *  If there is a syntax error in the selection string, then a
 *  runtime_error() is thrown.
*/

class Parser {
  loos::Kernel krnl;
  ParserDriver driver;

public:
  Parser(const string& s) : driver(s, krnl) { }

  //! Return a ref to the compiled (hopefully) Kernel.
  Kernel& kernel(void) { return(krnl); }
};





#endif

