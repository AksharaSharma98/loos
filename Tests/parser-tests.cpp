/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/




#include <iostream>
#include <vector>
#include <stdexcept>

#include "Atom.hpp"

#include "Kernel.hpp"
#include "Parser.hpp"

#include "Selectors.hpp"
#include "pdb.hpp"

using namespace std;
using namespace loos;

void test(PDB& pdb, const string s, bool expected = false) {
  Parser p;
  try {
    cout << "\n--------------------------------------\n";
    cout << "Parsing '" << s << "'\n";
    p.parse(s);
  }
  catch (...) {
    if (expected)
      cout << "Expected exception caught.\n";
    else
      cout << "===============================> UNEXPECTED EXCEPTION\n";
    return;
  }
  if (expected) {
    cout << "===============================> EXPECTED EXCEPTION NOT FOUND\n";
    return;
  }
  cout << p.kernel() << endl;
  KernelSelector ksel(p.kernel());
  AtomicGroup grp = pdb.select(ksel);
  cout << "Selected " << grp.size() << " @ " << grp.centroid();
  vector<GCoord> box = grp.boundingBox();
  cout << " & " << box[0] << " x " << box[1] << endl;
}


int main(int argc, char *argv[]) {

  PDB pdb(argv[1]);


  test(pdb, "name == 'CA'");
  test(pdb, "resid =~ '1\\d+'", true);
  test(pdb, "!(name == 'CA')");
  test(pdb, "!(name == 'CA'", true);
  test(pdb, "segid -> 'L(\\d+)' < 3");
  test(pdb, "(segid -> '(L|P)(\\d+)') <= 3");
  test(pdb, "(segid -> '(L|P)(\\d+)') <= 10 && name =~ 'C'");
  test(pdb, "name =~ 'C' && (resid >= 10 && resid <= 63) && segid != 'SOLV'");
  test(pdb, "!(name =~ 'C' && (resid >= 10 && resid <= 63) && segid != 'SOLV')");
}


