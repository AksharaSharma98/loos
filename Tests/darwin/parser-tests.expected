
--------------------------------------
Parsing 'name == 'CA''
Commands:
pushAtomName
pushString(<VALUE TYPE='STRING'>CA</VALUE>)
==


Selected 28 @ (29.5157,71.1618,51.4522) & (12.069,63.203,42.07) x (43.116,78.268,56.835)

--------------------------------------
Parsing 'resid =~ '1\d+''
Expected exception caught.

--------------------------------------
Parsing '!(name == 'CA')'
Commands:
pushAtomName
pushString(<VALUE TYPE='STRING'>CA</VALUE>)
==
!


Selected 497 @ (13.3716,43.8001,13.9885) & (-30.599,-33.181,-38.314) x (46.108,82.251,60.197)

--------------------------------------
Parsing '!(name == 'CA''
Expected exception caught.

--------------------------------------
Parsing 'hydrogen'
Commands:
Hydrogen


Selected 182 @ (2.11614,24.7385,-11.0071) & (-30.599,-33.181,-38.314) x (33.505,39.06,41.025)

--------------------------------------
Parsing '!hydrogen'
Commands:
Hydrogen
!


Selected 343 @ (20.6618,56.148,30.3097) & (-30.118,-32.397,-38.105) x (46.108,82.251,60.197)

--------------------------------------
Parsing 'resname == 'TIP3' && !hydrogen'
Commands:
pushAtomResname
pushString(<VALUE TYPE='STRING'>TIP3</VALUE>)
==
Hydrogen
!
&&


Selected 15 @ (1.208,-2.20973,5.77073) & (-30.118,-32.397,-38.105) x (32.577,36.303,41.516)

--------------------------------------
Parsing 'segid == 'BULK' && !(name =~ '^H')'
Commands:
pushAtomSegid
pushString(<VALUE TYPE='STRING'>BULK</VALUE>)
==
pushAtomName
matchRegex(^H)
!
&&


***ERROR***  Bad selection syntax - syntax error
***ERROR***  Bad selection syntax - syntax error
